flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraSinkBase.java

```diff
@@ -43,70 +48,82 @@
-	protected transient FutureCallback<V> callback;
+	// ------------------------ Default Configurations ------------------------
+
+	/**
+	 * The default maximum number of concurrent requests. By default, {@code Integer.MAX_VALUE}.
+	 */
+	public static final int DEFAULT_MAX_CONCURRENT_REQUESTS = Integer.MAX_VALUE;
+
+	/**
+	 * The default timeout duration when acquiring a permit to execute. By default, {@code Long.MAX_VALUE}.
+	 */
+	public static final long DEFAULT_MAX_CONCURRENT_REQUESTS_TIMEOUT = Long.MAX_VALUE;
+
+	/**
+	 * The default timeout unit when acquiring a permit to execute. By default, milliseconds.
+	 */
+	public static final TimeUnit DEFAULT_MAX_CONCURRENT_REQUESTS_TIMEOUT_UNIT = TimeUnit.MILLISECONDS;
+
+	// ------------------------- Configuration Fields -------------------------
+
+	private int maxConcurrentRequests = DEFAULT_MAX_CONCURRENT_REQUESTS;
+	private long maxConcurrentRequestsTimeout = DEFAULT_MAX_CONCURRENT_REQUESTS_TIMEOUT;
+	private TimeUnit maxConcurrentRequestsTimeoutUnit = DEFAULT_MAX_CONCURRENT_REQUESTS_TIMEOUT_UNIT;
+
+	// --------------------------- Cassandra Fields ---------------------------
 
 	private final ClusterBuilder builder;
 
-	private final AtomicInteger updatesPending = new AtomicInteger();
+	protected transient Cluster cluster;
+	protected transient Session session;
+
+	// ------------------------ Synchronization Fields ------------------------
+
+	private AtomicReference<Throwable> throwable;
+	private Semaphore semaphore;
+	private Phaser phaser;
 
 	CassandraSinkBase(ClusterBuilder builder) {
 		this.builder = builder;
 		ClosureCleaner.clean(builder, true);
 	}
 
-	@Override
-	public void open(Configuration configuration) {
-		this.callback = new FutureCallback<V>() {
-			@Override
-			public void onSuccess(V ignored) {
-				int pending = updatesPending.decrementAndGet();
-				if (pending == 0) {
-					synchronized (updatesPending) {
-						updatesPending.notifyAll();
-					}
-				}
-			}
+	// ----------------------------- Sink Methods -----------------------------
 
+	@Override
+	public void open(Configuration parameters) {
+		cluster = createCluster();
+		session = createSession();
+
+		throwable = new AtomicReference<>();
+		semaphore = new Semaphore(maxConcurrentRequests);
+		/*
+		 * A Phaser is a flexible and reusable synchronization barrier similar to CyclicBarrier and CountDownLatch.
+		 *
+		 * This Phaser is configured to support "1 + N" parties.
+		 *   - "1" for the CassandraSinkBase to arrive and to await at the Phaser during a flush() call.
+		 *   - "N" for the varying number of invoke() calls that register and de-register with the Phaser.
+		 *
+		 * The Phaser awaits the completion of the advancement of a phase prior to returning from a register() call.
+		 * This behavior ensures that no backlogged invoke() calls register to execute while the Semaphore's permits
+		 * are being released during a flush() call.
+		 */
+		phaser = new Phaser(1) {
```

Comment:
That's true; a concurrent set of `Future`s would be simpler. I've seen implementations of similar code in other OSS that rely on a concurrent `TrieMap` to track all the `pendingFutures` before a snapshot. At the snapshot, it iterates the `pendingFutures` calling `get()` to block which either returns or throws an exception. However, the `Future`s are created in large batches with Spark, so this implementation may have caveats in the streaming nature of Flink.

In Java, if we use a set backed by a `ConcurrentHashMap`, since we rely on a semaphore to throttle, we have deterministic behaviour regarding memory consumption, and since we can set the `concurrencyLevel` of a `ConcurrentHashMap` based on the underlying driver's parallelism, we should also have minimal contention on `put()`. The only feature we would miss given the streaming and back-pressured nature of Flink, is the Phaser preventing invoke() calls backed up on the semaphore from delaying the `flush()`. By introducing a flush lock with the `ConcurrentHashMap` and enough testing, it should be feasible to replace the `Phaser`. What do you think?

To your last comment, I originally relied on `Phaser` because the code originally used `wait` and `notify`, and I really like what Effective Java recommends: "Item 69: Prefer concurrency utilities to wait and notify". I was thinking of moving this logic into an abstract component, since Elasticsearch, Kafka, and so many other Sinks replicate a variant of the `wait` and `notify` logic, but I thought it was beyond scope. I was going to finish this ticket and have a follow-up suggesting we generalize this logic for all other sinks.

Reply:
If I understand correctly:
> The only feature we would miss given the streaming and back-pressured nature of Flink, is the Phaser preventing invoke() calls backed up on the semaphore from delaying the flush()

I think `invoke()` is not supposed to be called concurrently with `snapshotState()/flush()`or after `close()`.

Anyways, I see some space advantage in using `Phaser`.

Reply:
Actually, as calls to `invoke()`, `snapshotState()` and `close()` are serialised by checkpoint lock. It seems good enough to have only semaphore. When `flush()` is called, no new requests are created, so `flush()` could just wait for pending requests to finish by acquiring all the permits and then releasing them for next requests. What do you think?

Reply:
I believe the correct semantics for `flush()` is to behave as a dynamic, reusable barrier, and a Phaser by definition is that. By initializing a Phaser to 1, we initialize the CassandraSink with a barrier of size 1. As requests register and arrive and deregister with the Phaser, the dynamic size of the barrier changes accordingly. Thus, during `flush()`, when CassandraSink needs to wait for all the in-flight registered requests to finish, all it needs to do is arrive at the barrier and trigger it with an advance to wait for all the registered parties/requests arrive at the barrier as well.

Any attempt at implementing this behavior with Semaphores feels "hacky" to me because you would need to synchronize the Semaphore or some counter to find out how many permits were already acquired and try to acquire that number yourself. But doing that is just a variant of the discouraged practice of what was done before with wait-and-notify. Finally, several Semaphores are semantically used to implement barriers, so I find it bad to try to manually implement a non-standard barrier implementations with Semaphores when a barrier implementation is available.

Reply:
@jparkie How would the behavior differ if `flush()` would only call
```semaphor.acquireUninterruptibly(config.getMaxConcurrentRequests());```?

Reply:
For the `close()` scenario when the user specifies a `maxConcurrentRequestsTimeout`, if `invoke()` interleaves with `close()` such that the following operations happen: A2, A1, B1, C1, B2, C2. C2 would timeout and throw an exception. To prevent this, I would need to synchronize `close()` and `invoke()` together such that their method body's cannot interleave. Is that the only option?

```java
private volatile boolean isClosed;

@Override
public void close() {
  isClosed = true; // A1
  checkAsyncErrors(); // B1
  flush(); // C1
  checkAsyncErrors(); // D1
}

@Override
public void invoke(IN value) {
  if (isClosed) { // A2
    return;
  }
  checkAsyncErrors(); // B2
  tryAcquire(); // C2
  final ListenableFuture<V> result = send(value); // D2
  Futures.addCallback(result, new FutureCallback<V>() { // E2
    @Override
    public void onSuccess(V ignored) {
      release();
    }

    @Override
    public void onFailure(Throwable currentError) {
      setAsyncErrors(currentError);
      release();
    }
  });
}
```

Reply:
Good point.
We can release all in `flush` after  having done `acquireUninterruptibly`.
This should eventually unblock `tryAcquire`.
If timeout happens, I think, we should re-throw only if `isClosed` is false.
 
Would it help actually to check `isClosed` after `tryAcquire` in `invoke`?
Then either `invoke` proceeds and `flush` will wait for it or `flush` releases all and `invoke` can abort in case `isClosed` is true.

Reply:
Wait. It should throw an exception on timeout during closing right? If we short circuit early, without actually saving the data, it would be invalid. Is it okay to throw an exception on invoke() during closing?

Reply:
Thanks. Yeah. I was confused how other connectors handled the problem, but I thought they relied on the underlying drivers' (Kafka/Elasticsearch/...) `close()` worked.